### ISSUE TITLE
Authentication - handling the case where the user declines the FB permission request

### ISSUE THREAD

**Issue by robertlagrant** [SammyK/LaravelFacebookSdk#19]
Everything's running smoothly (thank you!). I'd like to polish things a bit by catching the case where the user declines the permission in Facebook. Is there a separate callback URL I can specify for this case?

Thanks!

**Comment by SammyK** [comment]
Hey @robertlagrant! Glad to hear things are running smoothly for you! :)

In the case a user removes a specific permission from your app, to my knowledge, there isn't a callback feature for that. You'd have to check the [`/{user-id}/permissions` endpoint](https://developers.facebook.com/docs/graph-api/reference/v2.2/user/permissions/) to see what permissions the user has authorized for your app. It's usually a good idea to check this endpoint right after a user authorizes your app to ensure mission-critical permissions were authorized.

If a user decides to deauthorize your whole app, Facebook has a "[Deauthorize Callback URL](https://developers.facebook.com/docs/facebook-login/manually-build-a-login-flow/v2.2#deauth-callback)" that you can configure in your app settings.

Hope that helps! Good luck!

**Comment by robertlagrant** [comment]
Thanks, I actually meant in the initial prompt when they sign up, but I'm just assuming that if the token is empty then they deliberately clicked Cancel on the Facebook screen.

**Comment by SammyK** [comment]
Ah yes. In the case the user denies the request, `getAccessToken()` will return `null` and not throw any exceptions.

Additionally, there will be several error params returned to the callback URL: `error`, `error_code`, `error_reason`, & `error_description`. v4.1 of the PHP SDK comes with [getter methods](https://github.com/facebook/facebook-php-sdk-v4/pull/276) for these on the `Facebook\Helpers\FacebookRedirectLoginHelper`:

```php
// Not logged in - user rejected or some other reason
if ( ! isset($accessToken)) {
  echo "<p>Error: " . $helper->getError() . "</p>\n\n";
  echo "<p>Error Code: " . $helper->getErrorCode() . "</p>\n\n";
  echo "<p>Error Reason: " . $helper->getErrorReason() . "</p>\n\n";
  echo "<p>Error Description: " . $helper->getErrorDescription() . "</p>\n\n";
  exit;
}
```

**Comment by robertlagrant** [comment]
Awesome, thanks. I'm getting some weird behaviour, particularly (for some reason; may be coincidental) with people on Android, and that may help diagnose.

**Comment by SammyK** [comment]
In the example above, the `$helper` object is an instantiation of `Facebook\Helpers\FacebookRedirectLoginHelper` which is what is used to try to obtain an access token from a redirect using `getAccessToken()`. The `getError()` getter methods are only available in v4.1. If you're using LaravelFacebookSdk v2.0 (in dev mode until SDK v4.1 release), you can obtain the redirect helper like so:

```php
$fb = App::make('laravel-facebook-sdk');
$helper = $fb->getRedirectLoginHelper();
```

Although this is subject to change. I've got some refactors in store for v2. :)

Good luck diagnosing the issue!
